text/3487-feature-visibility.md

```diff
@@ -0,0 +1,160 @@
+flag defaulting to `true` that indicates whether or not downstream crates should
+be allowed to use this feature.
+
+```toml
+[features]
+foo = { enables = [], public = false}
+```
+
+Attempting to use a private feature on a downstream crate will result in
+messages like the following:
+
+```
+error: feature `baz` on crate `mycrate` is private and cannot be used by
+  downstream crates
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+`public` is a boolean value that defaults to `true`. It can be thought of as
+`pub` in Rust source files, with the exception of being true by default. If set
+to `false`, Cargo should forbid its use with an error message on any downstream
+crates.
+
+The default `true` is not consistent with [`public_private_dependencies`] or
+Rust's `pub`, but is a reasonable default to be consistent with the current
+behavior. This means that either `feature = []` or
+`feature = { "enables" = [] }` will result in the same configuration.
+
+The name `public` was chosen in favor of `pub` to be consistent with the
+[`public_private_dependencies`] RFC, and to match the existing style of using
+non-truncated words as keys.
+
+In general, marking a feature `public = false` should make tooling treat the
+feature as non-public API. This is described as the following:
+
+- The feature is always usable within the same crate:
+  - Enabled by other features, e.g.
+    `foo = { enables = [some-private-feature] }`, is allowed
+  - Referenced in `[[bench]]` and `[[test]]` target `required-features`
+  - Using the feature on the command-line is allowed
+- Users may explicitly specifying the private features for their dependencies
+  on the command-line (e.g. `--features somecrate/private-feature`) which would
+  otherwise be forbidden
+- The feature should not be accepted by `cargo add --features`
+- The feature should not be reported from `cargo add`'s feature output report
+  - A future tool like `cargo info` shouldn't display information about these
+    features
+- Once `rustdoc` is able to consume feature metadata, `rustdoc` should not
+  document these features unless `--document-private-items` is specified
+
+Attempting to use a private feature in any of the forbidden cases should result
+in an error. Exact details of how features work will likely be refined during
+implementation and experimentation.
+
+Two sample use cases for `public = false` include:
+
+- `docs.rs` having a way to know which features should be hidden
+- Features that are included in feature chains (feature `a` enables feature `b`)
+  but not meant for public consumption could be marked not public
+
+This feature requires adjustments to the index for full support. This RFC
+proposes that it would be acceptable for the first implementation to simply
+strip private features from the manifest; this meanss that there will be no way
+to `cfg` based on these features.
```

Comment:
If we strip them, that affects a couple use cases
- When helping with managing the feature graph, you might want to cfg based on this
- Activating `debug` features in `clap` or `winnow` via `--features winnow/debug`

Reply:
In that case, do you think it is worth including the index change here?

Reply:
The index change can be fairly minor
- Add a new `private-features = [<name>, ...]` entry
  - `private` because those should be in the minority and keeps the index smaller
- Only allow them to be used when MSRV is high enough
  - Question is what to do when MSRV is unset.
  - It might be of interest for us to add other MSRV fields, like `published-rust-version` to record what version of rust was `cargo publish` done in and a `calculated-rust-version` that isn't as precise as `rust-version` but is a minimum value based on things like setting the edition or using private features
  - Alternatively, we could just make it a warning: "Hey, your rust-version is unset, if you publish with private features, older versions of cargo can use them and then will. break on upgrade"